### PR TITLE
Data structures update for LLT signals case

### PR DESCRIPTION
**Pull request by kkacprzak**
For v7.
Extension of data object classes with RecoFlag, that can be set during reconstruction to Good, Bad and Unknown (default). Adding necessary tests and changes in other classes, iterating the ClassDef versions. 
Also adding EventType::Cosmic, because why not.

### CODE REVIEW

File: Core/JPetManager/JPetManagerTest.cpp [hunk 1]
```diff
@@ -61,6 +61,7 @@ BOOST_AUTO_TEST_CASE(goodRootRun)
   };
   BOOST_REQUIRE_NO_THROW(manager.run(7, args));
 }
+*/
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply Core/JPetManager/JPetManagerTest.cpp]: yes, the file with old BaseSignals is not able to be procedded

File: DataObjects/JPetBaseSignal/JPetBaseSignalLinkDef.h [hunk 2]
```diff
@@ -1,9 +0,0 @@
-#ifdef __CINT__
-
-#pragma link off all globals;
-#pragma link off all classes;
-#pragma link off all functions;
-#pragma link C++ nestedclasses;
-#pragma link C++ class JPetBaseSignal+;
-
-#endif
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetBaseSignal/JPetBaseSignalLinkDef.h]: ok, I restored the LinkDefs
Just for check: other classes do not need LinkDefs?

File: DataObjects/JPetEvent/JPetEvent.h [hunk 3]
```diff
@@ -43,11 +43,15 @@ class JPetEvent: public TObject
 {
 
 public:
+  enum RecoFlag { Good, Bad, Unknown };
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetEvent/JPetEvent.h]: done

File: DataObjects/JPetLOR/JPetLOR.h [hunk 4]
```diff
@@ -35,10 +35,14 @@ class JPetHit;
 class JPetLOR: public TObject
 {
 public:
+  enum RecoFlag { Good, Bad, Unknown };
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetLOR/JPetLOR.h]: done

File: DataObjects/JPetPhysSignal/JPetPhysSignalTest.cpp [hunk 5]
```diff
@@ -46,7 +46,7 @@ BOOST_AUTO_TEST_CASE(ScalarFieldsTest)
 BOOST_AUTO_TEST_CASE(AllSignalsTest1)
 {
   // prepare raw signal
-  JPetRawSignal raw(4);
+  JPetRawSignal raw;
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetPhysSignal/JPetPhysSignalTest.cpp]: restored

File: DataObjects/JPetRawSignal/JPetRawSignal.cpp [hunk 6]
```diff
@@ -22,11 +22,7 @@ ClassImp(JPetRawSignal);
  *
  * @param Maximal number of points probed at one edge of the signal (defalult = 4)
  */
-JPetRawSignal::JPetRawSignal(const int points)
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetRawSignal/JPetRawSignal.cpp]: done

File: DataObjects/JPetSigCh/JPetSigCh.cpp [hunk 7]
```diff
@@ -14,55 +14,247 @@
...
-  fPM = NULL;
-  fFEB = NULL;
-  fTRB = NULL;
-  fTOMBChannel = NULL;
-}
+/**
+ * Default constructor
+ */
+JPetSigCh::JPetSigCh(): TObject(), fFlag(JPetSigCh::Unknown), fType(JPetSigCh::Leading),
+fValue(kUnset), fThreshold(kUnset), fThresholdNumber(0), fDAQch(0) {}
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetSigCh/JPetSigCh.cpp]: done, also in other places

File: DataObjects/JPetSigCh/JPetSigCh.cpp [hunk 8]
```diff
@@ -14,55 +14,247 @@
...
  * Constructor
- * @todo Perform some sanity checks of the given values
  */
-JPetSigCh::JPetSigCh(EdgeType Edge, float EdgeTime) {
-  init();
-  assert(EdgeTime > 0.);
-  fType = Edge;
-  fValue = EdgeTime;
+JPetSigCh::JPetSigCh(EdgeType edge, float time): TObject(), fFlag(JPetSigCh::Unknown),
+fType(edge), fValue(time), fThreshold(kUnset), fThresholdNumber(0), fDAQch(0) {}
```
[the remark this answers is not in the document]
Comment (kkacprzak) [reply DataObjects/JPetSigCh/JPetSigCh.cpp]: done

**Comment by kkacprzak** [comment]
Suggestions applied

File: Core/JPetManager/JPetManagerTest.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (alekgajos) [reply Core/JPetManager/JPetManagerTest.cpp]: OK, let's then leave this test commented and I will fix the test file afterwards.

File: Core/JPetManager/JPetManagerTest.cpp [hunk 9]
```diff
@@ -16,12 +16,11 @@
 #define BOOST_TEST_DYN_LINK
 #define BOOST_TEST_MODULE JPetManagerTest
 
-#include <boost/test/unit_test.hpp>
 #include "./JPetManager/JPetManager.h"
-#include <cstdio> //for std::remove
+#include <boost/test/unit_test.hpp>
+#include <cstdio>
```
Comment (wkrzemien) [inline Core/JPetManager/JPetManagerTest.cpp]: just the side note. Is cstdio include is used now?

File: DataObjects/JPetEventType/JPetEventType.h [hunk 10]
```diff
@@ -21,6 +21,7 @@ enum JPetEventType {
   k2Gamma = 2,
   k3Gamma = 4,
   kPrompt = 8,
-  kScattered = 16
+  kScattered = 16,
+  kCosmic = 32
```
Comment (wkrzemien) [inline DataObjects/JPetEventType/JPetEventType.h]: why do we need this for LLT?

File: DataObjects/JPetSigCh/JPetSigCh.cpp [hunk 11]
```diff
@@ -14,55 +14,247 @@
...
+{
+  fTOMBChannel = const_cast<JPetTOMBChannel*>(&channel);
+}
+
+/**
+ * Compares two SigChs by their threshold value
+ */
+bool JPetSigCh::compareByThresholdValue(const JPetSigCh& sigA, const JPetSigCh& sigB)
+{
+  if (sigA.getThreshold() < sigB.getThreshold()) {
```
Comment (wkrzemien) [inline DataObjects/JPetSigCh/JPetSigCh.cpp]: maybe just return (sigA.getThreshold() < sigB.getThreshold());

File: DataObjects/JPetSigCh/JPetSigCh.h [hunk 12]
```diff
@@ -27,176 +27,60 @@
...
-  float fThreshold; ///< value of threshold [mV]
-  int fDAQch; ///< Number of DAQ channel from the raw HLD file
-  TRef fPM;
-  TRef fFEB;
-  TRef fTRB;
-  TRef fTOMBChannel;
-
-  void init();
+private:
+  RecoFlag fFlag;
```
Comment (wkrzemien) [inline DataObjects/JPetSigCh/JPetSigCh.h]: you can set the default value here fFlag = ...
The same in fType

File: DataObjects/JPetEventType/JPetEventType.h (hunk 10 above)
Comment (kkacprzak) [reply DataObjects/JPetEventType/JPetEventType.h]: this is by the way, it will be needed for Cosmic Stream I guess

Comment (wkrzemien) [reply DataObjects/JPetEventType/JPetEventType.h]: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

File: Core/JPetManager/JPetManagerTest.cpp (hunk 9 above)
Comment (kkacprzak) [reply Core/JPetManager/JPetManagerTest.cpp]: it is not needed, removed

File: DataObjects/JPetSigCh/JPetSigCh.cpp (hunk 11 above)
Comment (kkacprzak) [reply DataObjects/JPetSigCh/JPetSigCh.cpp]: done

File: DataObjects/JPetSigCh/JPetSigCh.h (hunk 12 above)
Comment (kkacprzak) [reply DataObjects/JPetSigCh/JPetSigCh.h]: added here and in the other classes with new RecoFlag

**Comment by kkacprzak** [comment]
ad. 1 yes, Alek declared to modify the test files after
ad. 2. if there are no more changes requested, I can make such a test and post it here or somewhere else
ad. 3. I guess that is our approach right now, but maybe there can be other application of flags. I think it would be better to set those flags during the reconstruction/analysis in an agreed manner, leaving it in the core framework without forcing any behavior

**Comment by kkacprzak** [comment]
Performing size tests now. But I have one more question, that I just thought about:
should we change in the getter methods for TRefs the class casting from C-like to dynamic casts?
I.e. in JPetSigCh.cpp (line 93)
`const JPetPM & JPetSigCh::getPM() const {
  if(fPM.GetObject()) {
    return (JPetPM&) *fPM.GetObject();
  } else {
    ERROR("No JPetPM slot set, Null object will be returned");
    return JPetPM::getDummyResult();
  }
}`

change `(JPetPM&) *fPM.GetObject()` to `dynamic_cast<JPetPM&>(*fPM.GetObject())`
???

**Comment by wkrzemien** [comment]
Just not to lose time. I am going to accept it now. If you could send results later that would be great.

**Comment by alekgajos** [comment]
Concerning the C-style casts: we should change it at some point but I would not consider it a priority thing.
Certainly not before v7:-)

**Pull request merged**